crates/refunder/src/submitter.rs

```diff
@@ -0,0 +1,225 @@
+            tracing::warn!(
+                "Refunding txs are not started, as the current gas price {:?} \
+                            is higher than MAX_GAS_PRICE specified {:?}",
+                max_fee_per_gas,
+                MAX_GAS_PRICE
+            );
+            return Ok(());
+        }
+        self.gas_price_of_last_submission = Some(max_fee_per_gas);
+        self.nonce_of_last_submission = Some(nonce);
+        let tx_result = self
+            .ethflow_contract
+            .invalidate_orders_ignoring_not_allowed(encoded_ethflow_orders)
+            .gas_price(gas_price)
+            .from(self.account.clone())
+            .nonce(nonce)
+            .into_inner()
+            .resolve(resolve_conditions)
+            .send()
+            .await;
+        match tx_result {
+            Ok(handle) => {
+                tracing::debug!(
+                    "Tx to refund the orderuids {:?} yielded following result {:?}",
+                    uids,
+                    handle
+                );
+            }
+            Err(err) => {
+                let err = err.to_string();
+                if TX_ALREADY_MINED.iter().any(|msg| err.contains(msg)) {
+                    // It could happen that the previous tx got mined right before the tx was
+                    // send.
+                    tracing::debug!(?err, "transaction already mined");
+                } else if TX_ALREADY_KNOWN.iter().any(|msg| err.contains(msg)) {
+                    // This case means that the node is already aware of the tx
+                    // This can only happen after restarts, as usually we would always increase
+                    // the gas price compared to previous tx.
+                    // In this situation, we restart the submission loop with double the
+                    // previous gas. Due to the nature of EIP1559 this will not increase the
+                    // cost of the tx
+                    // todo: Finding last tx and its gas price would be nicer
+                    self.gas_price_of_last_submission = Some(max_fee_per_gas * 2);
+                    tracing::debug!(?err, "transaction already known");
+                } else {
+                    tracing::warn!(?err, "submission failed");
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+fn calculate_submission_gas_price(
+    gas_price_of_last_submission: Option<U256>,
+    web3_gas_estimation: GasPrice1559,
+    newest_nonce: U256,
+    nonce_of_last_submission: Option<U256>,
+) -> Result<GasPrice> {
+    let GasPrice1559 {
+        max_fee_per_gas, ..
+    }: gas_estimation::GasPrice1559 = web3_gas_estimation;
+    // The gas price of the refund tx is the current prevailing gas price
+    // of the web3 gas estimation plus a buffer.
+    // Since we are using Eip1559 gas specification,
+    // we will only pay the buffer if it is used
+    let mut new_max_fee_per_gas = max_fee_per_gas * GAS_PRICE_BUFFER_FACTOR;
```

Comment:
Why don't we do it this way in services too? I feel like we could simplify our code in the same way.